pool/swap.gno

```diff
@@ -0,0 +1,513 @@
+package pool
+
+import (
+	"std"
+
+	"gno.land/p/demo/ufmt"
+
+	"gno.land/r/gnoswap/v1/common"
+	"gno.land/r/gnoswap/v1/consts"
+
+	plp "gno.land/p/gnoswap/pool" // pool package
+
+	i256 "gno.land/p/gnoswap/int256"
+	u256 "gno.land/p/gnoswap/uint256"
+)
+
+// Swap swaps token0 for token1, or token1 for token0
+// Returns swapped amount0, amount1 in string
+// ref: https://docs.gnoswap.io/contracts/pool/pool.gno#swap
+func Swap(
+	token0Path string,
+	token1Path string,
+	fee uint32,
+	recipient std.Address,
+	zeroForOne bool,
+	amountSpecified string,
+	sqrtPriceLimitX96 string,
+	payer std.Address, // router
+) (string, string) {
+	common.IsHalted()
+	if common.GetLimitCaller() {
+		caller := std.PrevRealm().Addr()
+		if err := common.RouterOnly(caller); err != nil {
+			panic(addDetailToError(
+				errNoPermission,
+				ufmt.Sprintf("pool.gno__Swap() || only router(%s) can call pool swap(), called from %s", consts.ROUTER_ADDR, caller.String()),
+			))
+		}
+	}
+
+	if amountSpecified == "0" {
+		panic(addDetailToError(
+			errInvalidSwapAmount,
+			ufmt.Sprintf("pool.gno__Swap() || amountSpecified == 0"),
+		))
+	}
+
+	pool := GetPool(token0Path, token1Path, fee)
+
+	slot0Start := pool.slot0
+	if !slot0Start.unlocked {
+		panic(errLockedPool)
+	}
+
+	slot0Start.unlocked = false
+	defer func() { slot0Start.unlocked = true }()
+
+	amounts := i256.MustFromDecimal(amountSpecified)
+	sqrtPriceLimit := u256.MustFromDecimal(sqrtPriceLimitX96)
+
+	validatePriceLimits(pool, zeroForOne, sqrtPriceLimit)
+
+	feeGrowthGlobalX128 := getFeeGrowthGlobal(pool, zeroForOne)
+	feeProtocol := getFeeProtocol(slot0Start, zeroForOne)
+	cache := newSwapCache(feeProtocol, pool.liquidity)
+
+	state := newSwapState(amounts, feeGrowthGlobalX128, cache.liquidityStart, pool.slot0)
```

Comment:
```suggestion
	state := newSwapState(amounts, feeGrowthGlobalX128, cache.liquidityStart, slot0Start)
```

Because you are copying pool.slot0 to slot0Start, you must use slot0Start.
You should not access pool.slot0 directly.